### PR TITLE
Sprint/rearrange prepexp

### PR DESCRIPTION
**Pull request by mandresm**
Add a better description here. I just opened the PR to start to see if the tests pass or I broke something...

### CODE REVIEW

**Comment by nwieters** [comment]
Hi @mandresm , 
in order to set the variable `isresubmitted` in the new function `init_interactive_info` correctly, the jobtype (argument `-t`) needs to be given in the second call of esm_runscripts in prepexp. This is because the config entry for jobtype is set in `assemble_workflow` which will be run later on in prepare. Otherwise the entry config["general"]["jobtype"] is always unknown and therefore `isresubmitted` always `False` at the time `init_interactive_info` is called.

I will add this argument to the subprocess (previously os.system) when refactoring the 'copy_tools_to_thisrun' function.

**Comment by mandresm** [comment]
> the jobtype (argument -t) needs to be given in the second call of esm_runscripts in prepexp

I thought that was the case already, but you are right, it does not have a -t. If I add `--verbose` to the esm_runscripts command I can see:
```
cd /Users/mandresm/Work/deleteme2/scripts; esm_runscripts awicm3-v3.1-albedo-TCO95L91-CORE2.yaml -e deleteme2 --open-run -c --verbose --debug --no-motd  --last-jobtype prepcompute 
```

> I will add this argument to the subprocess (previously os.system) when refactoring the 'copy_tools_to_thisrun' function.

Okay thanks :)

**Comment by nwieters** [comment]
> I thought that was the case already, but you are right, it does not have a -t. If I add `--verbose` to the esm_runscripts command I can see:
> 
> ```
> cd /Users/mandresm/Work/deleteme2/scripts; esm_runscripts awicm3-v3.1-albedo-TCO95L91-CORE2.yaml -e deleteme2 --open-run -c --verbose --debug --no-motd  --last-jobtype prepcompute 
> ```
> 
The -t flag is added to esm_runscripts call in the *.run file already.

File: configs/esm_software/esm_runscripts/esm_plugins.yaml [hunk 1]
```diff
@@ -23,8 +34,9 @@ core:
                         - "_create_component_folders"
                         - "initialize_experiment_logfile"
                         - "copy_tools_to_thisrun"
+                        - "call_esm_runscripts_from_prepexp"
```
Comment (mandresm) [inline configs/esm_software/esm_runscripts/esm_plugins.yaml:37]:
```suggestion
                        - "call_esm_runscripts_from_prepexp"
                        # If the step above actually calls esm_runscripts (e.g. esm_runscripts called from 
                        # the experiment folder yet), the next steps will be skipped
```

File: src/esm_runscripts/prepexp.py [hunk 2]
```diff
@@ -1,25 +1,41 @@
...
+    config : dict
+        Dictionary containing the configuration information.
+    """
     evaluate(config, "prepexp", "prepexp_recipe")
     return config
 
 
 def color_diff(diff):
+    """
+
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py:37]:
```suggestion
    Adds color to text from a diff:
    * Green for lines starting with ``+``
    * Red for lines starting with ``-``
    * Blue for lines starting with ``^``

```

File: src/esm_runscripts/prepexp.py [hunk 3]
```diff
@@ -1,25 +1,41 @@
...
+    Parameters
+    ----------
+    diff :
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]:
```suggestion
    diff : str
        String to be colored
```

File: src/esm_runscripts/prepexp.py [hunk 4]
```diff
@@ -101,23 +118,13 @@ def copy_tools_to_thisrun(config):
...
-            print("Tools were updated, restarting...")
+    # If ``fromdir`` and ``scriptsdir`` are the same (the same as ``isresubmitted=True``),
+    # this is already a computing simulation which means we want to use the script
+    # in the experiment folder, so no copying is needed.
 
-        # At this point, ``fromdir`` and ``scriptsdir`` are different. Update the
-        # runscript if necessary
+    if not gconfig["isresubmitted"]:
+        # At this point, ``fromdir`` and ``scriptsdir`` are different (same as gconfig["isresubmitted"]=False).
+        # Update the runscript if necessary
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]:
```suggestion
```

Am I wrong by thinking this comment is incorrect? We can reach this condition by submitting `esm_runscripts` manually from the `script` directory in the experiment correct?

File: src/esm_runscripts/prepexp.py [hunk 5]
```diff
@@ -138,17 +145,94 @@ def copy_tools_to_thisrun(config):
...
+        Dictionary containing the configuration information.
+
+    """
+
+    gconfig = config["general"]
+
+    fromdir = os.path.realpath(gconfig["started_from"])
+    scriptsdir = os.path.realpath(gconfig["experiment_scripts_dir"])
+
+    # Return if already called from the experiment folder
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]:
```suggestion
    # Return if already called from the experiment folder without update flag
```

File: src/esm_runscripts/prepexp.py [hunk 6]
```diff
@@ -138,17 +145,94 @@ def copy_tools_to_thisrun(config):
...
+def call_esm_runscripts_from_prepexp(config):
+    """
+    Recipe step that creates a esm_runscripts command and submits this
+    to the functions that executes this command in a subprocess call.
+
+    Parameters
+    ----------
+    config : dict
+        Dictionary containing the configuration information.
+
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py:217]:
```suggestion

Returns
--------
config : dict
    Dictionary containing the configuration information.
```

File: src/esm_runscripts/prepexp.py [hunk 7]
```diff
@@ -138,17 +145,94 @@ def copy_tools_to_thisrun(config):
...
+    if config["general"]["verbose"]:
+        print(command)
+
+    # Exit after resubmission of esm_runscripts
+    end_it_all(config)
+
+def call_esm_runscripts_from_prepexp(config):
+    """
+    Recipe step that creates a esm_runscripts command and submits this
+    to the functions that executes this command in a subprocess call.
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]: Needs a comment about the fact that it doesn't resubmit if in experiment dir

File: src/esm_runscripts/prepexp.py [hunk 8]
```diff
@@ -374,27 +468,46 @@ def update_runscript(fromdir, scriptsdir, tfile, gconfig, file_type):
...
+    - Copies the setup *.date file from <experiment>/scripts/ folder
+      to <experiment>/run_xxxxxxxx-xxxxxxxx/scripts/ folder.
+    - Copies the runscript yaml file from current folder (<experiment>/scripts) 
+      to <experiment>/run_xxxxxxxx-xxxxxxxx/scripts/<runscript>
+    - Copies 'additional_files' (if any, e.g. fesom_output.yaml, that are called
+      via 'further_reading' in the runscript or other config file) from ...
+      to <experiment>/run_xxxxxxxx-xxxxxxxx/scripts/ folder.
+
+    Why here???
+
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]:
```suggestion
```

File: src/esm_runscripts/prepexp.py [hunk 9]
```diff
@@ -374,27 +468,46 @@ def update_runscript(fromdir, scriptsdir, tfile, gconfig, file_type):
...
     for filetype, filename, copy_or_link in filelist:
-        source = config["general"]["experiment_" + filetype + "_dir"]
-        dest = config["general"]["thisrun_" + filetype + "_dir"]
-        if copy_or_link == "copy":
-            method = shutil.copy2
-        elif copy_or_link == "link":
-            method = os.symlink
+        source = config["general"].get("experiment_" + filetype + "_dir", "")
+        dest = config["general"].get("thisrun_" + filetype + "_dir", "")
```
Comment (mandresm) [inline src/esm_runscripts/prepexp.py]:
```suggestion
        source = config["general"].get(f"experiment_{filetype}_dir", "")
        dest = config["general"].get(f"thisrun_{filetype}_dir", "")
```

**Review by mandresm — COMMENTED** [review]
Hi @nwieters, good job unknotting this part of the code. As you said, there might be some overlapping in the future with the method from the SimulationSetup that resubmits esm_runscripts commands if not in a cluster, but that is something we can simplify/reuse in the future. For now, once the suggestions are address, I think this can be merged to `release`

**Comment by nwieters** [comment]
Hi @mandresm ,
I applied your review suggestions and some additional format changes. Do you want to have a final review? 
I can't request a review from you, since this is your PR.
But if you have no further change requests I can approve this PR and merge it. Thanks.

**Comment by mandresm** [comment]
It all looks perfect, except for vilma-pism: https://github.com/esm-tools/esm_tools/actions/runs/8110920003/job/22169168392#step:8:656

Not particularly worried about vilma-pism but about the offline coupling. Vilma-pism won't run because it was depending on some environment files Dirk prepared on ollie and they are not in any other machine, however, it should be able to do a check (no compilation needed, only esm_runscripts with the line that is failing, runscript found in esm_tests/resources)

**Comment by mandresm** [comment]
Vilma-pism test is fixed. #approve-changes

**Comment by mandresm** [comment]
#bump #minor

**Pull request merged**